### ISSUE TITLE
Segmentation fault using python sdf reader

### ISSUE THREAD

**Issue by DoubleAgentDave** [epochpic/epoch#363]
I'm getting odd segmentation faults when trying to open distribution functions saved in epoch's sdf files. 

I've tried getting epoch to save all of the distribution functions to a single file, but then when opening them in python using the compiled sdf reader (just using d = sdf.read('xxx.sdf')) I get a segmentation fault.

When I've saved them along side other variables (e.g. fields) then it only issues a segmentation fault when I'm loading the distribution function. 

Oddly this behaviour is intermittent, and occasionally I can load a distribution function. 

When loading them through visit, this behaviour is also evident with the visit plugin either crashing, or just not loading that particular bit of data.

I've encountered errors loading variables before, but I can usually open them either with visit or python, but this time it's just (some) distribution functions and they don't seem to load anywhere.

I'm not exactly sure what details you'd need to start debugging this, I've no idea where to start myself.

**Comment by DoubleAgentDave** [comment]
I've managed to isolate it to only two cases, one where the distribution functions are output twice in two different output files (my bad, I had a duplicate entry in the input.deck), the other where the distribution functions were the only thing output with syntax in the input deck looking like this:

begin:output
  name = dists
  file_prefix = dists
  # number of timesteps between output dumps
  dt_snapshot = 10 * femto
  time_stop = 300*femto
  particles = never

  absorption = always
  total_energy_sum = always

  distribution_functions = always
end:output

in this case I can't access the distribution functions via the python sdf reader.

**Comment by DoubleAgentDave** [comment]
Just to make sure, I'm assuming the default behavior for 'distribution_functions' not being set is that they are not output at all in that particular dump?

**Comment by Quickrix** [comment]
Do you have dist_fn block(s) defined in your input.deck?

distribution_functions option has no sense without that, AFAK

**Comment by DoubleAgentDave** [comment]
Here is the syntax I'm using:

begin:output
  name = dists
  file_prefix = dists
  dt_snapshot = 10 * femto
  time_stop = 300*femto
  particles = never

  #simulation details
  absorption = always
  total_energy_sum = always
  
  distribution_functions = always
end:output

begin:dist_fn
  name = energy
  ndims = 1
  dumpmask = always
  direction1 = dir_energy
  range1 = (1, 1)
  resolution1 = 1000
  include_species:Electron
  include_species:Carbon
end:dist_fn

begin:dist_fn
  name = x_px
  ndims = 2
  dumpmask = always
  direction1 = dir_x
  direction2 = dir_px
  range1 = (1, 1)
  range2 = (1, 1)
  resolution1 = 1
  resolution2 = 500
  include_species:Electron
  include_species:Carbon
end:dist_fn

**Comment by DoubleAgentDave** [comment]
I've noticed that if I include the distribution functions with other output blocks, like electric fields or something then I can access them, it's just when they're output on their own, or just with scalar values.

**Comment by Quickrix** [comment]
Sorry, I failed to read your question carefully :)

It is probably a bug (that is out of my scope), but according to your description you can easily overcome it by adding any(?) "minor" data to your sdf with dist_fn output.

**Comment by DoubleAgentDave** [comment]
Yes I can do that, and currently they are being outputted with the rest of the grid variables. But for large scale 3D simulations there aren't small low cost outputs that can be made (besides the scalar variables described in the syntax above which didn't help). So if I need a distribution function with a higher (or lower) time series resolution than the grid/particle dumps it remains a problem. 

Unless the syntax above is somehow incorrect, which it might be :s

**Comment by erny123** [comment]
@DoubleAgentDave 

Thanks for posting. Running into the same issues.

**Comment by jthistle** [comment]
Possibly related to #693.

**Comment by TomGoffrey** [comment]
This is an old issue, but you can now install a standalone sdf reader via pip: https://pypi.org/project/sdfr/

If you encounter issues with this new package, please open a new issue.
